Automattic/jetpack
# [CRM] Add attribute definitions to workflows API results

Fixes https://github.com/Automattic/zero-bs-crm/issues/3225

## Proposed changes:


* Add `attribute_definitions` to all steps that includes attributes when exposed through any API endpoint.
* Define attributes for two actions (`Delete_Contact` and `Update_Contact_Status`).
  * Some conditions/actions have successfully defined attribute definitions but some of them are still missing. The two actions mentioned here are used in tests and lots of example code, so I went out of my way to make sure all of our previous shared snippets will continue to work. - Adding attribute definitions to the remaining conditions/actions is outside of scope for this PR.
* The Workflows API relies on actual Steps to exist so they can be hydrated, so a bug in `Automation_Faker` was discovered where the slug (`contact_status_condition`) did not match the actual mock condition (`contact_status`).
* Update the existing Workflow API tests to prune `attribute_definitions` when comparing hydrated (real) workflows with API responses.
* Update `Automation_Engine::get_registered_step` to properly hydrate a step like it was originally intended (but was never used before now).

### Other information:


## Jetpack product discussion



* pbhBOz-3Wf-p2
* p1695826185947649-slack-CTXBP902X

## Does this pull request change what data or activity we track or use?



No.

## Testing instructions:






### Automated tests

All of the tests should obviously pass, but looking at the following automated tests should suffice:

* `REST_Automation_Workflows_Controller_Test`: Updated so all the existing tests do not try to compare actual workflows with API responses that includes `Attribute_Definition`s.

### Manual testing

**Manual testing prep work**

1. Make sure the following feature flags are enabled on your site:

```php
add_filter( 'jetpack_crm_feature_flag_api_v4', '__return_true' );
add_filter( 'jetpack_crm_feature_flag_automations', '__return_true' );
```

2. Modify `REST_Automation_Workflows_Controller::get_item_permissions_check` to return `true` at the top.

_We do this to make it easier to call the endpoints. This should allow you to do it from any client you want (e.g. Postman) without having to worry about WP Logged in cookies, nonce, etc._

3. Make sure you have a workflow in your database. You can use the below snippet to add one.

```php
$workflow = new \Automattic\Jetpack\CRM\Automation\Automation_Workflow(
	[
		'name'         => 'New contact: Set status to "Customer"',
		'description'  => 'This automation will change the status of a contact to "Customer" when they are created.',
		'category'     => 'contact',
		'active'       => true,
		'triggers'     => [
			\Automattic\Jetpack\CRM\Automation\Triggers\Contact_Created::get_slug(),
		],
		'initial_step' => 1,
		'steps'        => [
			1 => [
				'slug'       => \Automattic\Jetpack\CRM\Automation\Actions\Update_Contact_Status::get_slug(),
				'attributes' => [
					'new_status' => 'Customer',
				],
				'next_step_true'  => null,
				'next_step_false' => null,
			],
		],
	]
);

$repo = new \Automattic\Jetpack\CRM\Automation\Workflow\Workflow_Repository();
$workflow = $repo->persist(  $workflow );
```

**GET All Workflows**

* Method: `GET`
* Path: `/wp-json/jetpack-crm/v4/automation/workflows`
* Payload: N/A

**GET Workflow**

* Method: `GET`
* Path: `/wp-json/jetpack-crm/v4/automation/workflows/{id}`
* Payload: N/A

## Workflow Response Example

```json
{
    "id": "1",
    "zbs_site": -1,
    "zbs_owner": -1,
    "name": "New contact: Set status to \"Customer\"",
    "description": "This automation will change the status of a contact to \"Customer\" when they are created.",
    "category": "contact",
    "triggers": [
        "jpcrm/contact_created"
    ],
    "steps": {
        "1": {
            "slug": "jpcrm/update_contact_status",
            "attributes": {
                "new_status": "Customer"
            },
            "next_step_true": null,
            "next_step_false": null,
            "attribute_definitions": {
                "new_status": {
                    "slug": "new_status",
                    "title": "New status",
                    "description": "The status that will be used for the contact.",
                    "type": "select",
                    "data": {
                        "Lead": "Lead",
                        "Customer": "Customer",
                        "Refused": "Refused",
                        "Blacklisted": "Blacklisted",
                        "Cancelled by Customer": "Cancelled by Customer",
                        "Cancelled by Us (Pre-Quote)": "Cancelled by Us (Pre-Quote)",
                        "Cancelled by Us (Post-Quote)": "Cancelled by Us (Post-Quote)"
                    }
                }
            }
        }
    },
    "initial_step": "1",
    "active": true,
    "version": 1,
    "created_at": 1695767341,
    "updated_at": 1695767341
}
```
